packages/firestore/src/remote/remote_store.ts

```diff
@@ -407,15 +426,43 @@ export class RemoteStore implements TargetMetadataProvider {
     }
 
     if (!snapshotVersion.isEqual(SnapshotVersion.min())) {
-      const lastRemoteSnapshotVersion = await this.localStore.getLastRemoteSnapshotVersion();
-      if (snapshotVersion.compareTo(lastRemoteSnapshotVersion) >= 0) {
-        // We have received a target change with a global snapshot if the snapshot
-        // version is not equal to SnapshotVersion.min().
-        await this.raiseWatchSnapshot(snapshotVersion);
+      try {
+        const lastRemoteSnapshotVersion = await this.localStore.getLastRemoteSnapshotVersion();
+        if (snapshotVersion.compareTo(lastRemoteSnapshotVersion) >= 0) {
+          // We have received a target change with a global snapshot if the snapshot
+          // version is not equal to SnapshotVersion.min().
+          await this.raiseWatchSnapshot(snapshotVersion);
+        }
+      } catch (e) {
+        logDebug(LOG_TAG, 'Failed to raise snapshot: ' + e);
+        await this.disableNetworkUntilRecovery(e);
       }
     }
   }
 
+  /**
+   * Recovery logic for IndexedDB errors that takes the network offline until
+   * IndexedDb probing succeeds. Retries are scheduled with backoff using
+   * `enqueueRetryable()`.
+   */
+  private async disableNetworkUntilRecovery(e: FirestoreError): Promise<void> {
+    if (e.name === 'IndexedDbTransactionError') {
+      ++this.indexedDbFailureBarrier;
```

Comment:
This mechanism seems more complicated/fiddly than it needs to be.

If we get multiple watch change events in a row, this is going to bump this `indexedDbFailureBarrier` multiple times and enqueue multiple retry tasks. When IndexedDB becomes available, each of them is going to succeed, decrement the count, and try to enable the network but only the last call to enableNetworkInternal will succeed.

It seems like instead of keeping a count, we could instead keep a boolean like `indexedDbFailed`. Then in this method:

  * If not `indexedDbFailed`, set it and enqueue the retry task
  * If `indexedDbFailed`, just log the error but do nothing.
  * In the retry task, if `getLastRemoteSnapshotVersion` succeeds, flip `indexedDbFailed` back to false and enable the network.

Reply:
You say "complicated/fiddly", I say "advanced" :) 

We should only ever get one of these events, since we go offline right away and discard all further messages from the stream. I realize that this is a good argument for making this a boolean and updated the code accordingly.